### PR TITLE
RavenDB-17977 - Support removing shard

### PR DESCRIPTION
**Pull request by lastav5**
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-17977

### Additional description

Removing an entire shard will happen when we remove all its shard databases (using `DeleteDatabaseCommand`), on the condition that there are no buckets mapped to it.
There is no dedicated EP for deleting an entire shard.

### Type of change

- New feature

### Backward compatibility

- Non breaking change

### Documentation update

- This change requires a documentation update. 

### Testing 

- Tests have been added that prove the fix is effective or that the feature works

### Is there any existing behavior change of other features due to this change?

- No

### UI work

- It requires further work in the Studio

### CODE REVIEW

File: src/Raven.Client/ServerWide/DatabaseRecord.cs [hunk 1]
```diff
@@ -468,6 +468,17 @@ internal static string GetKeyForDeletionInProgress(string node, int? shardNumber
 
             return $"{node}${shardNumber.Value}";
         }
+
+        public bool IsShardBeingDeletedOnAnyNode(int shardNumber)
+        {
+            foreach (var deletion in DeletionInProgress)
+            {
+                if (deletion.Key.Contains($"${shardNumber}"))
```
Comment (ayende) [inline src/Raven.Client/ServerWide/DatabaseRecord.cs]:
```suggestion
                if (deletion.Key.EndsWith($"${shardNumber}"))
```

File: src/Raven.Server/ServerWide/ClusterStateMachine.cs [hunk 2]
```diff
@@ -1472,54 +1472,61 @@ private void RemoveNodeFromDatabase(ClusterOperationContext context, BlittableJs
...
-                        rawRecord.Sharding.Shards.TryGetValue(shardNumber, out topology);
+                        rawRecord.Sharding.Shards.TryGetValue(shardNumber, out var topology);
+
+                        if (topology == null)
+                            return;
+
+                        databaseRecord = JsonDeserializationCluster.DatabaseRecord(rawRecord.Raw);
+                        remove.UpdateShardedDatabaseRecord(databaseRecord, shardNumber, index);
+
+                        if (databaseRecord.Sharding.Shards[shardNumber].Count == 0)
```
Comment (ayende) [inline src/Raven.Server/ServerWide/ClusterStateMachine.cs:1493]: `TryGetValue` here - what if we are trying to do that on a missing shard?

File: src/Raven.Server/ServerWide/ClusterStateMachine.cs [hunk 3]
```diff
@@ -1472,54 +1472,61 @@ private void RemoveNodeFromDatabase(ClusterOperationContext context, BlittableJs
...
                         remove.UpdateDatabaseRecord(databaseRecord, index);
-                    }
-                    else //shard
-                    {
-                        remove.UpdateShardedDatabaseRecord(databaseRecord, shardNumber, index);
-                    }
+                        
+                        if (databaseRecord.IsSharded)
+                        {
+                            throw new RachisApplyException($"Attempting to remove node from database {databaseName} but it is sharded.");
```
Comment (ayende) [inline src/Raven.Server/ServerWide/ClusterStateMachine.cs]: which node are you trying to remove?

File: src/Raven.Server/ServerWide/Commands/DeleteDatabaseCommand.cs [hunk 4]
```diff
@@ -60,11 +60,17 @@ public override void UpdateDatabaseRecord(DatabaseRecord record, long etag)
                             if (record.Sharding.Shards.ContainsKey(ShardNumber.Value) == false)
                                 throw new RachisApplyException($"The requested shard '{ShardNumber}' doesn't exists in '{record.DatabaseName}'");
 
+                            if (record.Sharding.Shards[ShardNumber.Value].ReplicationFactor == 1 && record.Sharding.DoesShardHaveBuckets(ShardNumber.Value))
```
Comment (ayende) [inline src/Raven.Server/ServerWide/Commands/DeleteDatabaseCommand.cs]: `TryGetValue`

File: src/Raven.Server/Web/System/AdminDatabasesHandler.cs [hunk 5]
```diff
@@ -716,34 +725,35 @@ public async Task Delete()
...
-                                }
+                                    if (topology.RelevantFor(node) == false)
+                                    {
+                                        throw new InvalidOperationException($"Database '{databaseName}' doesn't reside on node '{node}' so it can't be deleted from it");
+                                    }
+                                    
+                                    if (isShard && topology.ReplicationFactor == 1 && rawRecord.Sharding.DoesShardHaveBuckets(shardNumber))
+                                    {
+                                        throw new InvalidOperationException(
+                                            $"Database {databaseName} cannot be deleted because it is the last copy of shard {shardNumber} and it still contains buckets.");
```
Comment (ayende) [inline src/Raven.Server/Web/System/AdminDatabasesHandler.cs]: buckets may not tell enough to the user: "still contains data for the database" or something like that

File: src/Raven.Server/ServerWide/ClusterStateMachine.cs (hunk 2 above)
Comment (lastav5) [reply src/Raven.Server/ServerWide/ClusterStateMachine.cs:1493]: The `TryGetValue` a few lines above checks that

File: src/Raven.Server/ServerWide/Commands/DeleteDatabaseCommand.cs (hunk 4 above)
Comment (lastav5) [reply src/Raven.Server/ServerWide/Commands/DeleteDatabaseCommand.cs]: Also here, the lines above check `ContainsKey`

File: src/Raven.Server/ServerWide/ClusterStateMachine.cs (hunk 2 above)
Comment (ayende) [reply src/Raven.Server/ServerWide/ClusterStateMachine.cs:1493]: Then why are you calling this again? Use `topology` instead, no?

File: src/Raven.Server/ServerWide/Commands/DeleteDatabaseCommand.cs (hunk 4 above)
Comment (ayende) [reply src/Raven.Server/ServerWide/Commands/DeleteDatabaseCommand.cs]: Use `TryGetValue` to replace both checks

File: src/Raven.Server/ServerWide/ClusterStateMachine.cs (hunk 2 above)
Comment (lastav5) [reply src/Raven.Server/ServerWide/ClusterStateMachine.cs:1493]: Notice we deserialize the rawRecord (and also update it but there is no change happening to the Shards dict in that function). Do you think I should have another `TryGetValue` anyway?

File: src/Raven.Server/Web/System/AdminDatabasesHandler.cs (hunk 5 above)
Comment (lastav5) [reply src/Raven.Server/Web/System/AdminDatabasesHandler.cs]: Changed to "and it contains data that has not been migrated"

File: src/Raven.Server/ServerWide/Commands/DeleteDatabaseCommand.cs (hunk 4 above)
Comment (lastav5) [reply src/Raven.Server/ServerWide/Commands/DeleteDatabaseCommand.cs]: done

File: src/Raven.Server/ServerWide/ClusterStateMachine.cs (hunk 3 above)
Comment (lastav5) [reply src/Raven.Server/ServerWide/ClusterStateMachine.cs]: added

File: src/Raven.Client/ServerWide/DatabaseRecord.cs (hunk 1 above)
Comment (lastav5) [reply src/Raven.Client/ServerWide/DatabaseRecord.cs]: done

File: src/Raven.Client/ServerWide/Sharding/ShardingConfiguration.cs [hunk 6]
```diff
@@ -18,4 +18,20 @@ public class ShardingConfiguration
 
     // the dbid part with the MOVE tag upon migration
     public string DatabaseId;
+
+    public bool DoesShardHaveBuckets(int shardNumber) => DoesShardHaveBuckets(BucketRanges, shardNumber);
+
+    public static bool DoesShardHaveBuckets(List<ShardBucketRange> bucketRanges, int shardNumber)
+    {
+        if (bucketRanges == null)
```
Comment (karmeli87) [inline src/Raven.Client/ServerWide/Sharding/ShardingConfiguration.cs]: let's remove this check

File: src/Raven.Server/ServerWide/ClusterStateMachine.cs [hunk 7]
```diff
@@ -1472,54 +1472,69 @@ private void RemoveNodeFromDatabase(ClusterOperationContext context, BlittableJs
...
 
-                    DatabaseTopology topology;
-
-                    if (isShard)
+                    if (isShard) //shard database
                     {
-                        rawRecord.Sharding.Shards.TryGetValue(shardNumber, out topology);
+                        rawRecord.Sharding.Shards.TryGetValue(shardNumber, out var topology);
+
+                        if (topology == null)
```
Comment (karmeli87) [inline src/Raven.Server/ServerWide/ClusterStateMachine.cs]: better to 
```
if(rawRecord.Sharding.Shards.TryGetValue(shardNumber, out var topology) == false)
   return;
```

Comment (karmeli87) [reply src/Raven.Server/ServerWide/ClusterStateMachine.cs]: BTW where do you use the `topology` ?

File: src/Raven.Server/Web/System/AdminDatabasesHandler.cs [hunk 8]
```diff
@@ -690,11 +690,20 @@ public async Task Delete()
                     foreach (var databaseName in parameters.DatabaseNames)
                     {
                         DatabaseTopology topology = null;
-                        using (var rawRecord = ServerStore.Cluster.ReadRawDatabaseRecord(context, databaseName))
+                        var isShard = ShardHelper.TryGetShardNumberAndDatabaseName(databaseName, out string shardedDatabaseName, out int shardNumber);
```
Comment (karmeli87) [inline src/Raven.Server/Web/System/AdminDatabasesHandler.cs:693]: fine, 
but please open an issue to make a class or method out of this...
too many `ifs` for me :)

File: src/Raven.Server/ServerWide/ClusterStateMachine.cs (hunk 7 above)
Comment (lastav5) [reply src/Raven.Server/ServerWide/ClusterStateMachine.cs]: No where. Old code leftover. I guess It's enough to check `ContainsKey` here

File: src/Raven.Client/ServerWide/Sharding/ShardingConfiguration.cs (hunk 6 above)
Comment (lastav5) [reply src/Raven.Client/ServerWide/Sharding/ShardingConfiguration.cs]: done

File: src/Raven.Client/ServerWide/DatabaseRecord.cs [hunk 9]
```diff
@@ -468,6 +468,17 @@ internal static string GetKeyForDeletionInProgress(string node, int? shardNumber
 
             return $"{node}${shardNumber.Value}";
         }
+
+        public bool IsShardBeingDeletedOnAnyNode(int shardNumber)
```
Comment (ppekrol) [inline src/Raven.Client/ServerWide/DatabaseRecord.cs:472]: internal?

File: src/Raven.Client/ServerWide/DatabaseRecord.cs [hunk 10]
```diff
@@ -468,6 +468,17 @@ internal static string GetKeyForDeletionInProgress(string node, int? shardNumber
...
+        {
+            foreach (var deletion in DeletionInProgress)
+            {
+                if (deletion.Key.EndsWith($"${shardNumber}"))
```
Comment (ppekrol) [inline src/Raven.Client/ServerWide/DatabaseRecord.cs:476]: what is the Key here?

File: src/Raven.Client/ServerWide/Sharding/ShardingConfiguration.cs [hunk 11]
```diff
@@ -18,4 +18,17 @@ public class ShardingConfiguration
 
     // the dbid part with the MOVE tag upon migration
     public string DatabaseId;
+
+    public bool DoesShardHaveBuckets(int shardNumber) => DoesShardHaveBuckets(BucketRanges, shardNumber);
```
Comment (ppekrol) [inline src/Raven.Client/ServerWide/Sharding/ShardingConfiguration.cs:22]: internal

File: src/Raven.Client/ServerWide/Sharding/ShardingConfiguration.cs [hunk 12]
```diff
@@ -18,4 +18,17 @@ public class ShardingConfiguration
...
+
+    public static bool DoesShardHaveBuckets(List<ShardBucketRange> bucketRanges, int shardNumber)
```
Comment (ppekrol) [inline src/Raven.Client/ServerWide/Sharding/ShardingConfiguration.cs:24]: internal

File: src/Raven.Client/ServerWide/DatabaseRecord.cs (hunk 9 above)
Comment (lastav5) [reply src/Raven.Client/ServerWide/DatabaseRecord.cs:472]: done

File: src/Raven.Client/ServerWide/Sharding/ShardingConfiguration.cs (hunk 11 above)
Comment (lastav5) [reply src/Raven.Client/ServerWide/Sharding/ShardingConfiguration.cs:22]: done

File: src/Raven.Client/ServerWide/Sharding/ShardingConfiguration.cs (hunk 12 above)
Comment (lastav5) [reply src/Raven.Client/ServerWide/Sharding/ShardingConfiguration.cs:24]: done

File: src/Raven.Client/ServerWide/DatabaseRecord.cs (hunk 10 above)
Comment (lastav5) [reply src/Raven.Client/ServerWide/DatabaseRecord.cs:476]: of type {node}${shardNumber}

File: src/Raven.Client/ServerWide/DatabaseRecord.cs (hunk 9 above)
Comment (lastav5) [reply src/Raven.Client/ServerWide/DatabaseRecord.cs:472]: in https://github.com/ravendb/ravendb/pull/15577

**Pull request merged**